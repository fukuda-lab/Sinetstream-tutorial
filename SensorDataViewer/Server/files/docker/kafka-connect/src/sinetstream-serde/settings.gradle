rootProject.name = 'sinetstream-kafka-serde'
