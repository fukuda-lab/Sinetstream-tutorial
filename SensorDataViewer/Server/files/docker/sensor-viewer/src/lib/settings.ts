import { format, parseISO, sub } from 'date-fns';
import { persisted } from 'svelte-local-storage-store';
import {
  derived, writable, type Readable, type Writable,
} from 'svelte/store';
import * as zod from 'zod';

const paneVisibleSchema = zod.object({
  map: zod.boolean().optional(),
  image: zod.boolean().optional(),
  perftool: zod.boolean().optional(),
  lte: zod.boolean().optional(),
  sensors: zod.array(zod.string()).optional(),
});
const playerSchema = zod.object({
  from: zod.union([zod.preprocess(
    (arg) => (typeof arg === 'string' ? parseISO(arg) : arg),
    zod.date(),
  ), zod.string().regex(/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}$/)]),
  to: zod.union([zod.preprocess(
    (arg) => (typeof arg === 'string' ? parseISO(arg) : arg),
    zod.date(),
  ), zod.string().regex(/^\d{4}-\d{2}-\d{2} \d{2}:\d{2}$/)]),
  tick: zod.number().gte(100),
  speed: zod.number().gte(1).lte(100),
  repeat: zod.boolean(),
});
const chartV0Schema = zod.object({
  minSize: zod.object({
    width: zod.number().gte(200).optional(),
    height: zod.number().gte(150).optional(),
  }).optional(),
  line: zod.object({
    enabled: zod.boolean().optional(),
    size: zod.number().gt(0).optional(),
  }).optional(),
  point: zod.object({
    enabled: zod.boolean().optional(),
    size: zod.number().gt(0).optional(),
  }).optional(),
});
const layoutTemplateSchema = zod.enum(['horizontal', 'vertical-horizontal', 'vertical', 'vertical-2']);
const viewerSettingSchemaV0 = zod.object({
  name: zod.string(),
  publisher: zod.string().optional(),
  imageTopic: zod.string().optional(),
  perftoolName: zod.string().optional(),
  panes: paneVisibleSchema.optional(),
  imageUpdate: zod.enum(['push', 'polling']).optional(),
  player: playerSchema.optional(),
  layout: zod.object({ template: layoutTemplateSchema }).optional(),
  chart: chartV0Schema.optional(),
});
const androidSchema = zod.object({
  publisher: zod.string(),
  map: zod.boolean(),
  lte: zod.boolean(),
  sensors: zod.array(zod.string()),
});
const picameraSchema = zod.object({
  topic: zod.string(),
  update: zod.enum(['push', 'polling']),
  visible: zod.boolean(),
});
const layoutSchema = zod.object({
  template: layoutTemplateSchema,
});
const perftoolSchema = zod.object({
  name: zod.string(),
  visible: zod.boolean(),
});
const chartV1Schema = zod.object({
  minSize: zod.object({
    width: zod.number().gte(200),
    height: zod.number().gte(150),
  }),
  line: zod.object({
    enabled: zod.boolean(),
    size: zod.number().gt(0),
  }),
  point: zod.object({
    enabled: zod.boolean(),
    size: zod.number().gt(0),
  }),
});
const viewerSettingSchemaV1 = zod.object({
  name: zod.string(),
  version: zod.literal('1.0'),
  android: androidSchema,
  chart: chartV1Schema,
  image: picameraSchema,
  layout: layoutSchema,
  perftool: perftoolSchema,
  player: playerSchema,
});

type ViewerSettingV0 = zod.infer<typeof viewerSettingSchemaV0>;
type ViewerSettingV1 = zod.infer<typeof viewerSettingSchemaV1>;
type ViewerSetting = ViewerSettingV0 | ViewerSettingV1;
type PaneVisible = zod.infer<typeof paneVisibleSchema>;
type SensorViewerMode = 'viewer' | 'player';
type LayoutType = 'horizontal' | 'vertical-horizontal' | 'vertical' | 'vertical-2';
type ImageUpdateMethod = 'push' | 'polling';

function isViewerSettingV0(value: unknown): value is ViewerSettingV0 {
  return viewerSettingSchemaV0.safeParse(value).success;
}

function isViewerSettingV1(value: unknown): value is ViewerSettingV1 {
  return viewerSettingSchemaV1.safeParse(value).success;
}

const mode: Writable<SensorViewerMode> = persisted('mode', 'viewer');
const settings: Writable<ViewerSetting[]> = persisted('settings', []);
const settingsIndex = writable(0);

function defaultDateTime(minutes = 0): string {
  return format(sub(new Date(), { minutes }), 'yyyy-MM-dd HH:mm');
}

function toV1(v0: ViewerSettingV0): ViewerSettingV1 {
  return {
    name: v0.name,
    version: '1.0',
    android: {
      publisher: v0.publisher ?? '',
      map: v0.panes?.map ?? true,
      lte: v0.panes?.lte ?? true,
      sensors: v0.panes?.sensors ?? [],
    },
    chart: {
      minSize: {
        width: v0.chart?.minSize?.width ?? 310,
        height: v0.chart?.minSize?.height ?? 250,
      },
      line: {
        enabled: v0.chart?.line?.enabled ?? true,
        size: v0.chart?.line?.size ?? 1,
      },
      point: {
        enabled: v0.chart?.point?.enabled ?? true,
        size: v0.chart?.point?.size ?? 2,
      },
    },
    image: {
      topic: v0.imageTopic ?? '',
      update: v0.imageUpdate ?? 'push',
      visible: v0.panes?.image ?? true,
    },
    layout: {
      template: v0.layout?.template ?? 'horizontal',
    },
    perftool: {
      name: v0.perftoolName ?? '',
      visible: v0.panes?.perftool ?? true,
    },
    player: {
      from: v0.player?.from ?? defaultDateTime(15),
      to: v0.player?.to ?? defaultDateTime(),
      tick: v0.player?.tick ?? 1200,
      speed: v0.player?.tick ?? 10,
      repeat: v0.player?.repeat ?? false,
    },
  };
}

function defaultSetting(settingName: string): ViewerSettingV1 {
  return toV1({ name: settingName });
}

const currentSetting: Readable<ViewerSettingV1> = derived(
  [settings, settingsIndex],
  ([$settings, $settingsIndex], set) => {
    if ($settingsIndex < $settings.length && $settingsIndex >= 0) {
      const current = $settings[$settingsIndex];
      if (isViewerSettingV1(current)) {
        set(current);
      } else if (isViewerSettingV0(current)) {
        set(toV1(current));
      }
    } else {
      set(defaultSetting(''));
    }
  },
);

const name: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.name);
});

const publisher: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.android.publisher);
});

const perftoolName: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.perftool.name);
});

const imageTopic: Readable<string> = derived([currentSetting], ([$currentSetting], set) => {
  set($currentSetting.image.topic);
});

const pane: Readable<PaneVisible> = derived([currentSetting], ([$currentSetting], set) => {
  const current = $currentSetting;
  set({
    lte: current.android.lte,
    map: current.android.map,
    sensors: current.android.sensors,
    image: current.image.visible,
    perftool: current.perftool.visible,
  });
});

const charts: Readable<string[]> = derived([currentSetting], ([$currentSetting], set) => {
  const current = $currentSetting;
  const { sensors } = current.android;
  const result: string[] = [...sensors];
  if (current.android.lte) {
    result.unshift('lte');
  }
  if (current.perftool.visible) {
    result.unshift('perftool');
  }
  set(result);
});

const imageUpdateMethod: Readable<ImageUpdateMethod> = derived(
  [currentSetting],
  ([$currentSetting], set) => {
    set($currentSetting.image.update);
  },
);

export {
  settings,
  settingsIndex,
  mode,
  name,
  publisher,
  perftoolName,
  imageTopic,
  pane,
  charts,
  imageUpdateMethod,
  currentSetting,
  isViewerSettingV0,
  isViewerSettingV1,
  viewerSettingSchemaV1,
};
export type {
  SensorViewerMode,
  ImageUpdateMethod,
  ViewerSetting,
  ViewerSettingV0,
  ViewerSettingV1,
  PaneVisible,
  LayoutType,
};
