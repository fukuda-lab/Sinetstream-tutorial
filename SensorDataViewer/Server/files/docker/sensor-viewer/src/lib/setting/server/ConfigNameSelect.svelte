<script lang="ts">
  import { getContextClient, gql, queryStore } from "@urql/svelte";
  import { createEventDispatcher } from "svelte";
  import { currentSetting } from "../../settings";
  import ErrorToast from "./ErrorToast.svelte";

  const client = getContextClient();
  const queryNames = queryStore({
    client,
    query: gql`
      query {
        viewer_config {
          id
          name
        }
      }
    `,
  });
  const dispatch = createEventDispatcher();

  export let name: string;
  export let selected = -1;
  let container: any;

  $: if (container && $queryNames?.data && selected < 0) {
    const rst = $queryNames.data.viewer_config.filter(
      (item: Record<string, any>) => item.name === $currentSetting.name
    );
    if (rst.length > 0) {
      dispatch("match", { id: rst[0].id });
    }
  }
</script>

<select {name} bind:this={container} aria-label="config-name-select">
  {#if selected < 0}
    <option value="" selected>--Please choose an option--</option>
  {/if}
  {#if $queryNames?.data}
    {#each $queryNames.data.viewer_config as item}
      <option value={item.id}>{item.name}</option>
    {/each}
  {/if}
</select>
{#if $queryNames?.error}
  <ErrorToast message={$queryNames.error.message} />
{/if}
