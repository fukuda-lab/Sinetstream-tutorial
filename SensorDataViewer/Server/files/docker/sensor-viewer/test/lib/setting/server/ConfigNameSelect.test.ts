import {
  act, cleanup, getByText, render,
} from '@testing-library/svelte';
import * as urql from '@urql/svelte';
import { writable } from 'svelte/store';
import {
  afterEach, beforeEach, describe, expect, it, vi,
} from 'vitest';
import ConfigNameSelect from '../../../../src/lib/setting/server/ConfigNameSelect.svelte';
import { settings, settingsIndex } from '../../../../src/lib/settings';

describe('ConfigNameSelect.svelt', () => {
  beforeEach(() => {
    vi.mock('@urql/svelte');
    settings.set([{ name: 'setting-0' }]);
    settingsIndex.set(0);
  });

  afterEach(() => cleanup());

  it('no data', () => {
    const { container, getAllByRole } = render(ConfigNameSelect, { props: { name: 'name' } });
    const opts = getAllByRole('option');
    expect(opts).toHaveLength(1);
    expect(opts[0]).toHaveProperty('value', '');
    expect(container.outerHTML).matchSnapshot();
  });

  describe('two items', () => {
    let result;

    beforeEach(() => {
      result = writable({ fetching: true, data: undefined, error: undefined });
      const spy = vi.spyOn(urql, 'queryStore');
      spy.mockReturnValue(result);
    });

    it('no match', async () => {
      const { component, container, getAllByRole } = render(ConfigNameSelect, { props: { name: 'name' } });
      expect(container.outerHTML).toMatchSnapshot();
      const handler = vi.fn();
      component.$on('match', handler);

      await act(() => {
        result.update((x) => {
          const data = {
            viewer_config: [1, 2].map(
              (n) => ({ id: n, name: `setting-${n}` }),
            ),
          };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      const opts = getAllByRole('option');
      expect(opts).toHaveLength(3);
      expect(opts[0]).toHaveProperty('value', '');
      [1, 2].forEach((n) => {
        expect(opts[n]).toHaveProperty('value', n.toString());
        expect(getByText(opts[n], `setting-${n}`)).toBeTruthy();
      });
      expect(container.outerHTML).toMatchSnapshot();
      expect(handler).not.toHaveBeenCalled();
    });

    it('match event', async () => {
      settings.set([{ name: 'setting-2' }]);
      const { component, container, getAllByRole } = render(ConfigNameSelect, { props: { name: 'name' } });
      expect(container.outerHTML).toMatchSnapshot();
      const handler = vi.fn();
      component.$on('match', handler);

      await act(() => {
        result.update((x) => {
          const data = {
            viewer_config: [1, 2].map(
              (n) => ({ id: n, name: `setting-${n}` }),
            ),
          };
          const fetching = false;
          return { ...x, fetching, data };
        });
      });
      const opts = getAllByRole('option');
      expect(opts).toHaveLength(3);
      expect(opts[0]).toHaveProperty('value', '');
      [1, 2].forEach((n) => {
        expect(opts[n]).toHaveProperty('value', n.toString());
        expect(getByText(opts[n], `setting-${n}`)).toBeTruthy();
      });
      expect(container.outerHTML).toMatchSnapshot();
      expect(handler).toHaveBeenCalledOnce();
      expect(handler.mock.calls[0][0]?.detail).toMatchObject({ id: 2 });

      await act(() => {
        component.$set({ selected: 2 });
      });
      expect(getAllByRole('option')).toHaveLength(2);
      expect(container.outerHTML).toMatchSnapshot();
    });
  });

  it('error', async () => {
    const ERROR_MESSAGE = '[Network] Failed to fetch';
    const result = writable({ fetching: true, data: undefined, error: undefined });
    const spy = vi.spyOn(urql, 'queryStore');
    spy.mockReturnValue(result);

    const { container, getAllByRole } = render(ConfigNameSelect, { props: { name: 'name' } });
    expect(container.outerHTML).toMatchSnapshot();

    await act(() => {
      result.update((x) => {
        const error = { message: ERROR_MESSAGE };
        const fetching = false;
        return { ...x, fetching, error };
      });
    });
    const opts = getAllByRole('option');
    expect(opts).toHaveLength(1);
    expect(container.outerHTML).toMatchSnapshot();
    expect(getByText(container, ERROR_MESSAGE)).toBeTruthy();
  });
});
