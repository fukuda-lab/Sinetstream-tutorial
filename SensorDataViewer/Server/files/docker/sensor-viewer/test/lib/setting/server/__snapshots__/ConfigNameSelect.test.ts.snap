// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfigNameSelect.svelt > error 1`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"\\">--Please choose an option--</option></select> <!--<ConfigNameSelect>--></div></body>"`;

exports[`ConfigNameSelect.svelt > error 2`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"\\">--Please choose an option--</option></select> <div class=\\"toast pink white-text active top\\" aria-label=\\"error-toast\\"><i>error</i> <span>[Network] Failed to fetch</span></div><!--<ErrorToast>--><!--<ConfigNameSelect>--></div></body>"`;

exports[`ConfigNameSelect.svelt > no data 1`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"\\">--Please choose an option--</option></select> <!--<ConfigNameSelect>--></div></body>"`;

exports[`ConfigNameSelect.svelt > two items > match event 1`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"\\">--Please choose an option--</option></select> <!--<ConfigNameSelect>--></div></body>"`;

exports[`ConfigNameSelect.svelt > two items > match event 2`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"\\">--Please choose an option--</option><option value=\\"1\\">setting-1</option><option value=\\"2\\">setting-2</option></select> <!--<ConfigNameSelect>--></div></body>"`;

exports[`ConfigNameSelect.svelt > two items > match event 3`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"1\\">setting-1</option><option value=\\"2\\">setting-2</option></select> <!--<ConfigNameSelect>--></div></body>"`;

exports[`ConfigNameSelect.svelt > two items > no match 1`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"\\">--Please choose an option--</option></select> <!--<ConfigNameSelect>--></div></body>"`;

exports[`ConfigNameSelect.svelt > two items > no match 2`] = `"<body><div><select name=\\"name\\" aria-label=\\"config-name-select\\"><option value=\\"\\">--Please choose an option--</option><option value=\\"1\\">setting-1</option><option value=\\"2\\">setting-2</option></select> <!--<ConfigNameSelect>--></div></body>"`;
